### ISSUE TITLE
Rank Evaluation Metric NDCG inconsistent between LGBMRanker.fit vs. LGBMRanker.predict()

### ISSUE THREAD

**Issue by cjsombric** [microsoft/LightGBM#6814]
## Description
LGBMRanker's eval metric 'ndcg' does not match (1) the NDCG I generate by hand or (2) the NDCG generated by sklearn.metrics.ndcg_score.
I believe this is caused by different predictions being generated in the "fit" and "predict" LGBMRanker processes.

## Reproducible example
`
from lightgbm import LGBMRanker
import pandas as pd

# unable to include training data:
X_train = []
y_train = []
train_groups = []

# Will include a small sample validation set:
X_ex_small = [] # too big to include
relevance_data = [0, 1, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
y_ex_small = pd.Series(relevance_data)
y_ex_small.index = ['123'] * len(y_ex_small)

val_ex_small = pd.Series([20], index=["123"])

model_gdbt = LGBMRanker(objective="lambdarank",
                   boosting_type = "gbdt",
                   random_state=42,
                    max_depth=10,         
                    min_data_in_leaf=200, 
                    n_estimators=100,      
                    subsample=0.5,         
                    colsample_bytree=0.6, 
                    lambda_l1=0.9,
                    lambda_l2=0.9,
                   n_jobs = -1)

model_gdbt.fit(X_train, 
               y_train, 
               group=train_groups,
               eval_set=[(X_train, y_train), (X_ex_small, y_ex_small)],
               eval_group=[train_groups, val_ex_small],
               eval_metric=['ndcg'],
               eval_at=[1, 5])

# lightgbm's NDCG for the validation set (X_ex_small, y_ex_small)
lightgbm_ndcg5 =  model_gdbt.evals_result_['valid']['ndcg@5'][-1] # returns 0.9596358141506394

# sklearn NDCG for the validation set (X_ex_small, y_ex_small)
from sklearn.metrics import ndcg_score
y_ex_small_pred = model_gdbt.predict(X_ex_small) # returns [3.35057029, 3.46270607, 3.49636596, 3.13087368, 3.55823747, 3.28307494, 3.37924097, 2.84076022, 3.39463737, 3.51163148, 3.497507  , 3.29053029, 3.25150361, 3.12610702, 3.06495292, 2.70117685, 3.17200958, 3.52930069, 2.90185473, 3.51950751]
sklearn_ndcg5 =  ndcg_score([y_ex_small.values], [y_ex_small_pred], k=5) # returns 0.8637574337885663

# Manual NDCG for the validation set (X_ex_small, y_ex_small)

#Visulize the relevance scores (y_ex_small) nest to the predicted relevance scores (y_ex_small_pred) and sort by the predicted relevance
df_temp = y_ex_small.to_frame()
df_temp["pred"] = y_ex_small_pred
df_temp.sort_values(by="pred", ascending=False)

# returns:
#relevance | pred
#-- | --
#4 | 3.558237
#0 | 3.529301
#0 | 3.519508
#0 | 3.511631
#0 | 3.497507
#0 | 3.496366
#1 | 3.462706
#...

# Manual computation of ndcg@5 which matches sklearn, but not lightgbm
from math import log2
dcg5 = ( 4 / log2(1 + 1))
idcg5 = ( 4 / log2(1 + 1)) + ( 1 / log2(1 + 2))
manual_ndcg5 = dcg5/idcg5 # returns 0.8637574337885663

# Thus manual_ndcg5==sklearn_ndcg5 
# BUT manual_ndcg5!=lightgbm_ndcg5

# In order to replicate the lightgbm_ndcg5 I need to assume the predicted ranking had the relevance 1 document in the 4th position...
from math import log2
dcg5_2replicate_lightgbm_ndcg5 = ( 4 / log2(1 + 1)) + ( 1 / log2(1 + 4))
idcg5_2replicate_lightgbm_ndcg5 = ( 4 / log2(1 + 1)) + ( 1 / log2(1 + 2))
manual_ndcg5_2replicate_lightgbm_ndcg5 = dcg5_2replicate_lightgbm_ndcg5/idcg5_2replicate_lightgbm_ndcg5 # returns 0.9596358141506394

`

## Environment info

LightGBM version or commit hash:
LightGBM: 4.5.0
sklearn: 1.5.2

Command(s) you used to install LightGBM

```
pip install lightgbm

```

In summary, there are differences in the reported performance (ndcg) depending on which LGBMRanker functions you use. I believe this discrepancy is from differences in the predicted relevance scores generated in the LGBMRanker.fit() vs. LGBMRanker.predict() processes.

**Comment by cjsombric** [comment]
Here is a replication of the problem with synthetic data. Note (1) the manual calculation will have to be updated based on the observed outputs of the model and (2) you need multiple relevance score levels to replicate this error.

`# %% [markdown]
# ### Simple example to show lightGBM NDCG issue

# %%
from sklearn.datasets import make_classification
import numpy as np
from lightgbm import LGBMRanker
import random

# %% [markdown]
# ### Create Data

# %%
# Make a synthetic ranking dataset for demonstration
seed = 1994
X, y = make_classification(random_state=seed)
y[[random.randint(0, 99) for _ in range(10)]] = 4 # include 4 as a relevance score
rng = np.random.default_rng(seed)
n_query_groups = 3
qid = rng.integers(0, n_query_groups, size=X.shape[0])

# Sort the inputs based on query index
sorted_idx = np.argsort(qid)
X = X[sorted_idx, :]
y = y[sorted_idx]
qid = qid[sorted_idx]

# %%
import pandas as pd

df_X = pd.DataFrame(data=X,    # values
              index=qid,    # 1st column as index
              )
df_y = pd.DataFrame(data=y,    # values
              index=qid,    # 1st column as index
              )

train_group_size = df_X.index.value_counts()

# %%
# Make an example small, single query data set for evaluation with 10 documents
X_ex_small = X[[random.randint(0, 99) for _ in range(10)], :]
y_ex_small = np.array([0, 0, 1, 0, 1, 0, 0, 0, 4 , 0])
qid_ex_small = 3 * np.ones(10)

df_X_ex_small = pd.DataFrame(data=X_ex_small,    # values
              index=qid_ex_small,    # 1st column as index
              )
df_y_ex_small  = pd.DataFrame(data=y_ex_small,    # values
              index=qid_ex_small,    # 1st column as index
              )
evaluate_group_size = df_X_ex_small .index.value_counts()

# %% [markdown]
# ### Train a Model

# %%
model_gdbt = LGBMRanker(objective="lambdarank",
                   boosting_type = "gbdt",
                   random_state=42)

# %%
model_gdbt.fit(
            df_X, 
               df_y, 
               group=train_group_size,
               eval_set=[(df_X, df_y),  (df_X_ex_small, df_y_ex_small)],
               eval_group=[train_group_size,evaluate_group_size],
               eval_metric=['ndcg'],
               eval_at=[1, 5, 10]
          )

# %%
# print(model_gdbt.evals_result_["valid_1"]['ndcg@1'][-1])
print(model_gdbt.evals_result_["valid_1"]['ndcg@5'][-1])
# print(model_gdbt.evals_result_["valid_1"]['ndcg@10'][-1])

# %%
# When I generate the predictions using the predicted functionality of LGBMRanker.predict I get a different ndcg score
from sklearn.metrics import ndcg_score
y_ex_small_pred = model_gdbt.predict(df_X_ex_small)
ndcg_score([df_y_ex_small[0].values], [y_ex_small_pred], k=5)

# %%
# Visualizing the predicted relevance compared to actual relevance
df_temp = pd.DataFrame()
df_temp["true"] = df_y_ex_small
df_temp["pred"] = y_ex_small_pred
df_temp.sort_values(by="pred", ascending=False)

# %%
# Manually computing the =ndcg@5 for the small example query and the result agrees with sklearn
from math import log2

dcg5 = (1 / log2(1 + 3))
idcg5 = ( 1 / log2(1 + 1)) + ( 1 / log2(1 + 2)) + ( 1 / log2(1 + 3)) 


print(dcg5/idcg5)


`

**Comment by jameslamb** [comment]
Thanks, someone will answer here when they can.

I've reformatted your posts. You can click "edit" to see the raw text for how I did that. Before you post here again, please read https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

**Comment by jameslamb** [comment]
It looks like you posted a similar question on `xgboost`, linking that here for reviewers: https://github.com/dmlc/xgboost/issues/11235

**Comment by cjsombric** [comment]
@jameslamb Thank you for helping me improve! :+1:

**Comment by cjsombric** [comment]
I also checked using a different DCG and was not able to replicate the NDCG's from lightGBM

```python
# Manually computing the =ndcg@5 for the small example query with a stricter DCG

dcg5= ( (2**(4) -1) / log2(1 + 6)) + (  (2**(1) -1) / log2(1 + 2)) + (  (2**(1) -1) / log2(1 + 5)) 
idcg5 = (  (2**(4) -1) / log2(1 + 1)) + (  (2**(1) -1) / log2(1 + 2)) + (  (2**(1) -1) / log2(1 + 3)) 

print(dcg5/idcg5)
```
